src/Time/Units.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE TypeOperators      #-}
+
+module Time.Units
+       ( -- * Time
+         Time (..)
+
+         -- ** Time data types
+       , Second
+       , MilliSecond
+       , MicroSecond
+       , NanoSecond
+       , Minute
+       , Hour
+       , Day
+       , Week
+       , Fortnight
+
+         -- ** Units
+       , SecondUnit
+       , MilliSecondUnit
+       , MicroSecondUnit
+       , NanoSecondUnit
+       , MinuteUnit
+       , HourUnit
+       , DayUnit
+       , WeekUnit
+       , FortnightUnit
+       ) where
+
+import Data.Ratio (Rational)
+
+import Time.Rational (type (%), Rat)
+
+
+newtype Time (rat :: Rat) = Time Rational
```

Comment:
You can derive `Num` instance for this data type using `GeneralizedNewtypeDeriving` extension. This instance would be useful.

Also, I guess, you need to replace `Rational` with `Ratio Natural`. `Rational` is `Ratio Integer`. We use `GHC.TypeNats` module which allows to reflect `Nat` from type into `Natural` data type, not `Integer`. 

I asked a question about difference between `Natural` and `Integer` some time ago:

* https://stackoverflow.com/questions/45904898/when-to-pick-natural-over-integer-in-haskell

Reply:
Is negative time impossible in your model?

Reply:
@int-index I didn't find reasons and use-cases for having negative time.

Reply:
What if I use the `time` type to represent a timestamp starting from 0 AD and then I have a BC date?

Reply:
@int-index Are you aware of `Either` data type?

Reply:
Furthermore, I think there should be two types:

1. Timestamp — can be negative, has no units
2. Time (TimeDiff) — cannot be negative, has units

and instead of `Num` instances that provides subtraction, there must be operations like:

```
data DiffSign = Before | After

timediff :: KnownRat unit => Timestamp -> Timestamp -> (DiffSign, Time unit)
timeadd :: KnownRat unit => Time unit -> Time unit -> Time unit
timemult :: KnownRat unit => Ratio Natural -> Time unit -> Time unit
```

Reply:
>  Are you aware of Either data type?

I know I can add a sign externally, but this would mean I have to redefine arithmetic operations. Anyway, now that I think of it, arithmetic operations on time don't fit into the `Num` hierarchy anyway (see the comment above).

Reply:
To make my reasoning more intuitive, imagine a timeline. `Timestamp` is a point on this line, and `Time unit` is an interval measured in `unit`.

Reply:
@int-index What `Timestamp` should represent? Is it microseconds or nanoseconds? If `Timestamp` is `Integer` for example? What `Timestamp` stored internally?

Reply:
> What Timestamp should represent?

A point on the timeline.

> Is it microseconds or nanoseconds?

It's unitless.

> If Timestamp is Integer for example? What Timestamp stored internally?

Arbitrary precision, it's `Rational`.